pkg/operator/latencyprofilecontroller/profilerejectionchecker.go

```diff
@@ -0,0 +1,92 @@
+package latencyprofilecontroller
+
+import (
+	"fmt"
+
+	listersv1 "k8s.io/client-go/listers/core/v1"
+
+	configv1 "github.com/openshift/api/config/v1"
+	nodeobserver "github.com/openshift/library-go/pkg/operator/configobserver/node"
+)
+
+type profileRejectRevisionChecker struct {
+	configMapLister           listersv1.ConfigMapNamespaceLister
+	usedConfigPaths           [][]string
+	knownProfileConfigs       map[configv1.WorkerLatencyProfileType]map[string]interface{}
+	profileRejectionScenarios []nodeobserver.LatencyProfileRejectionScenario
+}
+
+// NewInstallerProfileRejectionChecker is used to construct a CheckProfileRejectionFunc that
+// can compare the (installer controller generated config map) config of the latest active revision
+// with the desired/target latency profile to return whether profile transition should be rejected or not.
+// The latencyConfigs passed to this func check arg values of each latency profile to compare
+// them across the active config on the cluster. In case the transition is from profile found in active config
+// to specified target profile and it is for one of the scenarios specified in profileRejectionScenarios,
+// the rejection flag is returned as true with an appropriate message; false otherwise.
+// Eg. transition between extreme latency profiles viz. Default <-> Low should be rejected.
+func NewInstallerProfileRejectionChecker(
+	configMapLister listersv1.ConfigMapNamespaceLister,
+	latencyConfigs []nodeobserver.LatencyConfigProfileTuple,
+	profileRejectionScenarios []nodeobserver.LatencyProfileRejectionScenario,
+) (CheckProfileRejectionFunc, error) {
+
+	knownProfileConfigs, err := nodeobserver.GenerateConfigsForEachLatencyProfile(latencyConfigs)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := profileRejectRevisionChecker{
+		configMapLister:           configMapLister,
+		usedConfigPaths:           nodeobserver.GetUsedLatencyConfigPaths(latencyConfigs),
+		knownProfileConfigs:       knownProfileConfigs,
+		profileRejectionScenarios: profileRejectionScenarios,
+	}
+	return ret.checkProfileRejection, nil
+}
+
+func (r *profileRejectRevisionChecker) checkProfileRejection(targetProfile configv1.WorkerLatencyProfileType, currentRevisions []int32) (isRejected bool, rejectMsg string, err error) {
+	// do not reject at day-0 and support any profile
+	if nodeobserver.IsDayZero(currentRevisions) {
+		return false, "", nil
+	}
+
+	// determine the highest revision among the ones specified, since the highest revision
+	// will be indicative of the most recent profile active on the cluster
+	highestCurrentRevision := currentRevisions[0]
```

Comment:
hm shouldn't we check all revisions for the profile transition rejection?

if you do `LowUpdateSlowReaction` -> `MediumUpdateAverageReaction` -> `Default` quickly you will get the right error, but once first node reaches `MediumUpdateAverageReaction` we will stop showing the error in favor of progressing, even though we are still requesting invalid profile.

Is this the behaviour we want?

Reply:
also progressing=False in this scenario I interpret as: we we are not progressing to your requested rejected profile, even though we are progressing to a profile in the middle. And we will eventually get also to your rejected profile and show progressing.

I wonder if it might be confusing?

Reply:
I guess that would be intended behaviour since our actual intent is not dis-allowing the user to update their cluster going from Low latency profile to Default latency profile and vice versa, but only ensure that during transition we do not break the cluster at any point in time. So rejection(s) in case of direct jump is exclusively in favour of cluster stability since we risk breaking the cluster if we jump rapidly from Low -> Default as also stated in the [EP](https://github.com/openshift/enhancements/blob/master/enhancements/worker-latency-profile/worker-latency-profile.md#lowupdateslowreaction---default) for which we ought to deal with the rejections.

I tested it out on a cluster, going from Low -> Medium -> Default in succession quickly doesn't seem to break anything (as long as when going from Low -> Medium had atleast 1 of the node's current revision has Medium profile active) and since we have the second suppressor (`suppressUntilSameConfigProfiles`) in place, it takes care that at any point in time not more than 2 profiles are revision active on the nodes as also discussed previously in https://github.com/openshift/library-go/pull/1364#discussion_r869878699 .

So, for an example scenario it would be like:
1. initially, `profile=Low; nodes={Low, Low, Low}; status=Complete`
2. user changes profile, `profile=Medium; nodes={Low, Low, Low}; status=Progresssing`
3. some progress is made `profile=Medium; nodes={Medium, Low, Low}; status=Progresssing`
4. user changes profile again, `profile=Default; nodes={Medium, Low, Low}; status=Progresssing`
5. second suppressor takes action, so observedConfig is still at older Medium, `profile=Default; nodes={Medium, Low, Low}; status=Progresssing`
6. ... after progress is made, `profile=Default; nodes={Medium, Medium, Medium}; status=Progresssing`
7. either suppressors don't suppress this time, observedConfig is updated to rollout Default, `profile=Default, nodes={Default, Medium, Medium}; status=Progression`
8. and finally, when all nodes reach Default, `profile=Default; nodes={Default, Default, Default}; status=Complete`.

This is the same reasoning why we introduced `highestCurrentRevision`, in order to help support such behaviour where we give users the ability to update profiles in succession without having to break the cluster. And, as discussed previously we could introduce automation for going from Default to Low and IMHO, we're providing something very close to that without breaking anything. Moreover, we can thoroughly document this with risk and advantages for awareness about it!

Reply:
> hm shouldn't we check all revisions for the profile transition rejection?

So checking for all possible current revisions has another subtle problem which I just recalled. If we are comparing profiles between more than one current revision, we can have a situation where revision-x is at Default and revision-y is at Medium and our target profile set on the cluster is Low, so we don't get consensus on whether to reject or not.